exchanges/okx/okx_wrapper.go

```diff
@@ -0,0 +1,1218 @@
+	})
+	if err != nil {
+		return err
+	}
+	er := ok.Websocket.SetupNewConnection(stream.ConnectionSetup{
+		URL:                  okxAPIWebsocketPublicURL,
+		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
+		ResponseMaxLimit:     okxWebsocketResponseMaxLimit,
+	})
+	if er != nil {
+		return er
+	}
+	return ok.Websocket.SetupNewConnection(stream.ConnectionSetup{
+		URL:                  okxAPIWebsocketPrivateURL,
+		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
+		ResponseMaxLimit:     okxWebsocketResponseMaxLimit,
+		Authenticated:        true,
+	})
+}
+
+// Start starts the Okx go routine
+func (ok *Okx) Start(wg *sync.WaitGroup) error {
+	if wg == nil {
+		return fmt.Errorf("%T %w", wg, common.ErrNilPointer)
+	}
+	wg.Add(1)
+	go func() {
+		ok.Run()
+		wg.Done()
+	}()
+	return nil
+}
+
+// Run implements the Okx wrapper
+func (ok *Okx) Run() {
+	if ok.Verbose {
+		log.Debugf(log.ExchangeSys,
+			"%s Websocket: %s.",
+			ok.Name,
+			common.IsEnabled(ok.Websocket.IsEnabled()))
+		ok.PrintEnabledPairs()
+	}
+	if !ok.GetEnabledFeatures().AutoPairUpdates {
+		return
+	}
+	err := ok.UpdateTradablePairs(context.TODO(), false)
+	if err != nil {
+		log.Errorf(log.ExchangeSys,
+			"%s failed to update tradable pairs. Err: %s",
+			ok.Name,
+			err)
+	}
+}
+
+// FetchTradablePairs returns a list of the exchanges tradable pairs
+func (ok *Okx) FetchTradablePairs(ctx context.Context, a asset.Item) ([]string, error) {
+	if !ok.SupportsAsset(a) {
+		return nil, fmt.Errorf("asset type of %s is not supported by %s", a, ok.Name)
+	}
+	format, err := ok.GetPairFormat(a, false)
+	if err != nil {
+		return nil, err
+	}
+	pairs := []string{}
+	insts := []Instrument{}
+	var er error
+	switch a {
+	case asset.Spot:
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeSpot,
+		})
+	case asset.Futures:
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeFutures,
+		})
+	case asset.PerpetualSwap:
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeSwap,
+		})
+	case asset.Option:
+		var instsb []Instrument
+		var instsc []Instrument
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "BTC-USD",
+		})
+		if er != nil {
+			goto checkErrorAndContinue
+		}
+		instsb, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "ETH-USD",
+		})
+		if er != nil {
+			goto checkErrorAndContinue
+		}
+		instsc, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "SOL-USD",
+		})
+		insts = append(insts, instsb...)
+		insts = append(insts, instsc...)
+	case asset.Margin:
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeMargin,
+		})
+	}
+checkErrorAndContinue:
+	if er != nil || len(insts) == 0 {
```

Comment:
separate errors as this can return nil and no values. Have a specific error for no instruments returned.